radix-engine/src/engine/state_track.rs

```diff
@@ -0,0 +1,239 @@
+use core::ops::RangeFull;
+
+use indexmap::{IndexMap, IndexSet};
+use sbor::rust::rc::Rc;
+use sbor::rust::vec::Vec;
+use scrypto::buffer::{scrypto_decode, scrypto_encode};
+
+use crate::engine::Address;
+use crate::ledger::*;
+
+use super::{
+    track::VirtualSubstateId, SubstateOperation, SubstateOperationsReceipt, SubstateParentId,
+    SubstateValue,
+};
+
+/// Keeps track of state changes that that are non-reversible, such as fee payments
+pub struct BaseStateTrack {
```

Comment:
There must be a way to integrate this with the current recursive track. Seems like both structures essentially do the same thing?

Reply:
Not sure if we need recursiveness in transactional state track. Also, the difference is it doesn't care about substate ID.